code/modules/error_handler/error_viewer.dm

```diff
@@ -0,0 +1,149 @@
+// Error viewing datums, responsible for storing error info, notifying admins
+// when errors occur, and showing them to admins on demand.
+
+// There are 3 different types used here:
+//
+// - ErrorCache keeps track of all error sources, as well as all individually
+//   logged errors. Only one instance of this datum should ever exist, and it's
+//   right here:
+
+#ifdef DEBUG
+var/global/datum/ErrorViewer/ErrorCache/error_cache = new()
+#else
+// If debugging is disabled, there's nothing useful to log, so don't bother.
+var/global/datum/ErrorViewer/ErrorCache/error_cache = null
+#endif
+
+// - ErrorSource datums exist for each line (of code) that generates an error, 
+//   and keep track of all errors generated by that line.
+//
+// - ErrorEntry datums exist for each logged error, and keep track of all
+//   relevant info about that error.
+
+// Common vars and procs are kept at the ErrorViewer level
+/datum/ErrorViewer/
+	var/name = ""
+
+/datum/ErrorViewer/proc/browseTo(var/user, var/html)
+	if(user)
+		user << browse(html, "window=error_viewer;size=600x400")
+
+/datum/ErrorViewer/proc/buildHeader(var/datum/ErrorViewer/back_to, var/linear)
+	// Common starter HTML for showTo
+	. = {"<html><head><title>Runtime Viewer</title>
+	<style>body{font-family:'Courier New',monospace;font-size:8pt;}</style>
+	</head><body>"}
+
+	if(istype(back_to))
+		. += "[back_to.makeLink("<<<", null, linear)]<br><br>"
+	return .
```

Comment:
well actually defining a hard `return .` defeats most of the point of using `.` as there is extra overhead to an early return compared to letting the function exit